### DISCUSSION TITLE
Dealing with Bad Photo Data - Strategies?

### DISCUSSION THREAD

**Discussion by tootbrute** [photoprism/photoprism#888] [General]
**Backstory**:
I've imported about 50,000 photos into Photoprism. Dates are the most important metadata for me. I have many files that have the incorrect Exif/date data embedded in the files. This means every time I reindex my files, I lose dates, I changed in the Photoprism database.

So... I want to fix these 'bad' file dates.

Suggestions for easiest way to do this?

My idea is this so far...
**Option 1**
1. Download files
2. Archive in Photoprism (Delete not supported in GUI yet*)
3 Fix with date with Exif tools
4. Reimport into Photoprism

In the future when batch editing is supported, I am also thinking of just tagging the bad files with "wrongdate" so I can deal them when I can.

_*I run Photoprism on a Linux server, so finding the files to delete them manually via CLI is quite a pain._

I think I need to wait for delete support, and batch editing support to even think of trying to tackle this problem though.

**Option 2**
Right now doing Google Photos takeout. I will see what kind of files are here. I might use the files, or just use the albums to jog my memory about the real dates for some of these trips.

=======

How are you guys dealing with bad Exif data?
Maybe I'm thinking about this the wrong way. Maybe I should adjust all dates in Photoprism, and then backup the database, and never do a complete reindex.

**Comment by andrewlow** [comment]
I did an initial bulk import of my tree of photo archives - this included iPhoto library, Photos library (both OSX based software for those that don't recognize them) - and a much older manually curated Photos-Backup.

This was ~80k photos, included many scans & old photos from old digital cameras with bad EXIF data. I selected to 'IMPORT' vs. 'Index' - this was intentional because I wanted to fix the bad organization I had and remove duplicates.

I ran photoprism in pure docker with the defaults. This resulted in ~15k+ in the 'review' state - and many, many with no EXIF data making it into the collection.

Search was busted (for me) because photos with no date sorted to the top - so before I could scroll through the most recent - I had to deal with all the old data. The UI doesn't yet support multi-editing EXIF data (https://github.com/photoprism/photoprism/issues/271) so my best approach was to set that effort aside and start again.

(and many kudo's to the photoprism team - my criticism above isn't intended to complain, but I want to lay out the pain I caused myself with a bad start - we collectively need to arrive at better advice for moving data into photoprism)

I'm on Linux as the hosting platform for photoprism - that's where all my storage is.

So far - my plan is to do more sifting through the photos I plan to import before shoving them all into PhotoPrism via import. The most important thing is to figure out which ones do not have EXIF data and avoid sending those in before I can fix them.

`exiftool` is the best story it seems. My plan is to set the file dates as correctly as I can and use exiftool to create EXIF data based on the file dates.

Here is my super simple bash script that helps find 'bad' files

```
#!/bin/bash


find . -type f -print0 | while read -d $'\0' file; do
        if exiftool -if '$datetimeoriginal' "$file" &> /dev/null; then
                printf 'OK %s\t' "$file"
                exiftool -time:datetimeoriginal "$file"
        else
                echo 'boom', $file
        fi
done
```

You can skip all of the 'OK' lines by piping the output of this script through `grep -v "^OK"` 

The good news - is that it seems large chunks of my archives have valid EXIF data - so I get get a lot of data into photoprism without pain. Then I need to spend more time scripting up what I'll do for the many that do not - but as I dig through my library I realize there is a lot of junk in there too. 

Hopefully this is helpful... but then again, my base advice is to fix the data that you put into PhotoPrism first :)

**Reply by andrewlow** [reply]
I found this https://ryanmo.co/2014/08/01/find-images-with-no-exif-dates/

Which got me better sorted out with exiftool capabilities.

Query all in this file tree - dump out 'bad' EXIF files
```
exiftool -filename -r -if '(not $datetimeoriginal or ($datetimeoriginal eq "0000:00:00 00:00:00"))' -common -csv .
```

Set a date - you can also do a full directory instead of the filename
```
exiftool -datetimeoriginal="2000:10:01 00:00:00" filename
```
For the brave that have already imported many into their photoprim library - you CAN modify the files in 'originals' if you are brave. I'm taking the more conservative approach and working my way through my source trees before copying into the import folder.
